File: examples/execution_control/operator_status_tracking/python/operator_status_tracking.py
"""
SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""  # noqa: E501

import time

from holoscan.conditions import CountCondition, PeriodicCondition
from holoscan.core import Application, Operator, OperatorSpec, OperatorStatus
from holoscan.schedulers import EventBasedScheduler


class FiniteSourceOp(Operator):
    """A simple operator that processes a fixed number of times and then completes."""

    def __init__(self, fragment, *args, max_count=5, **kwargs):
        super().__init__(fragment, *args, **kwargs)
        self.max_count = max_count
        self.count = 0

    def setup(self, spec: OperatorSpec):
        spec.output("out")

    def compute(self, op_input, op_output, context):
        if self.count < self.max_count:
            print(f"[{self.name}] Emitting data: {self.count}")
            op_output.emit(self.count, "out")

            # Simulate some processing time
            time.sleep(0.2)
        else:
            print(f"[{self.name}] Additional iterations after completion: {self.count}")

        self.count += 1

    def stop(self):
        print(f"[{self.name}] Stopping operator")


class ProcessorOp(Operator):
    """An operator that processes data."""

    def setup(self, spec: OperatorSpec):
        spec.input("in")
        spec.output("out")

    def compute(self, op_input, op_output, context):
        data = op_input.receive("in")

        # Process the data
        result = data * 2
        print(f"[{self.name}] Processing data: {data} -> {result}")

        # Emit the processed data
        op_output.emit(result, "out")

        # Simulate some processing time
        time.sleep(0.1)

    def stop(self):
        print(f"[{self.name}] Stopping operator")


class ConsumerOp(Operator):
    """An operator that consumes data."""

    def setup(self, spec: OperatorSpec):
        spec.input("in")

    def compute(self, op_input, op_output, context):
        data = op_input.receive("in")

        # Consume the data
        print(f"[{self.name}] Consuming data: {data}")

        # Simulate some processing time
        time.sleep(0.15)

    def stop(self):
        print(f"[{self.name}] Stopping operator")


class MonitorOp(Operator):
    """A dedicated monitoring operator that runs independently of the main pipeline."""

    def __init__(self, fragment, *args, **kwargs):
        super().__init__(fragment, *args, **kwargs)
        self.monitored_operators = kwargs.pop("monitored_operators", [])
        self.consecutive_idle_count = 0

    def compute(self, op_input, op_output, context):
        print(f"[{self.name}] Operator status summary:")

        is_pipeline_idle = True
        for op_name in self.monitored_operators:
            status = context.get_operator_status(op_name)
            print(f"  - {op_name}: {status.name}")
            if status != OperatorStatus.IDLE:
                is_pipeline_idle = False

        if is_pipeline_idle:
            self.consecutive_idle_count += 1
            print("!!IDLE!!")

            # Currently, there's no way to retrieve the computed SchedulingCondition status from the
            # operator (i.e., whether the computed scheduling condition status is NEVER).
            # Instead, we use `consecutive_idle_count` to determine if all operators, except the
            # monitor operator, have completed.
            # If `consecutive_idle_count` is equal to or greater than the hardcoded value 3,
            # we consider them completed, as there are cases where all operators are idle but not
            # yet completed.
            # A better approach for checking an operator's computed scheduling condition will be
            # available in a future release.
            if self.consecutive_idle_count >= 3:
                print(f"[{self.name}] All operators have completed.")
                # Stop the monitor operator which is the only operator keeping the application alive
                self.stop_execution()  # the application will terminate through a deadlock
        else:
            self.consecutive_idle_count = 0


class OperatorStatusTrackingApp(Application):
    def compose(self):
        # Define the operators
        source = FiniteSourceOp(self, CountCondition(self, 10), name="source", max_count=5)
        processor = ProcessorOp(self, name="processor")
        consumer = ConsumerOp(self, name="consumer")

        # Create the independent monitor operator with its own scheduling condition
        # This will continue running even if the main pipeline stops
        monitor = MonitorOp(
            self,
            PeriodicCondition(self, 0.05),  # 50ms in seconds
            name="monitor",
            monitored_operators=["source", "processor", "consumer"],
        )

        # Define the workflow for the main pipeline
        self.add_flow(source, processor, {("out", "in")})
        self.add_flow(processor, consumer, {("out", "in")})

        # The monitor is not connected to any other operators - it runs independently
        self.add_operator(monitor)

        # Print information about the execution context API
        print("This example demonstrates the Operator Status Tracking API.")
        print("The source operator will emit 5 values and then stop executing after 10 iterations.")
        print("The monitor operator runs independently and tracks the status of all operators.")
        print("When operators complete, the monitor will be terminated.")
        print("-------------------------------------------------------------------")


def main():
    app = OperatorStatusTrackingApp()
    scheduler = EventBasedScheduler(app, worker_thread_number=2)
    app.scheduler(scheduler)
    app.run()

    print("-------------------------------------------------------------------")
    print("Application completed. All operators have finished processing.")


if __name__ == "__main__":
    main()

File: examples/execution_control/operator_status_tracking/cpp/operator_status_tracking.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <chrono>
#include <iostream>
#include <string>
#include <thread>
#include <vector>

#include <holoscan/holoscan.hpp>
#include <magic_enum.hpp>

// A simple operator that processes a fixed number of times and then completes
class FiniteSourceOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(FiniteSourceOp)

  FiniteSourceOp() = default;

  void setup(holoscan::OperatorSpec& spec) override {
    spec.output<int>("out");
    spec.param(max_count_, "max_count", "Maximum Count", "Maximum number of times to emit data", 5);
  }

  void compute([[maybe_unused]] holoscan::InputContext& op_input,
               [[maybe_unused]] holoscan::OutputContext& op_output,
               [[maybe_unused]] holoscan::ExecutionContext& context) override {
    if (count_ < max_count_.get()) {
      std::cout << "[" << name() << "] Emitting data: " << count_ << std::endl;
      op_output.emit(count_, "out");

      // Simulate some processing time
      std::this_thread::sleep_for(std::chrono::milliseconds(200));
    } else {
      std::cout << "[" << name() << "] Additional iterations after completion: " << count_
                << std::endl;
    }

    count_++;
  }

  void stop() override { std::cout << "[" << name() << "] Stopping operator" << std::endl; }

 private:
  holoscan::Parameter<int> max_count_;
  int count_ = 0;
};

// An operator that processes data
class ProcessorOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(ProcessorOp)

  ProcessorOp() = default;

  void setup(holoscan::OperatorSpec& spec) override {
    spec.input<int>("in");
    spec.output<int>("out");
  }

  void compute(holoscan::InputContext& op_input, holoscan::OutputContext& op_output,
               [[maybe_unused]] holoscan::ExecutionContext& context) override {
    auto data = op_input.receive<int>("in").value();

    // Process the data
    int result = data * 2;
    std::cout << "[" << name() << "] Processing data: " << data << " -> " << result << std::endl;

    // Emit the processed data
    op_output.emit(result, "out");

    // Simulate some processing time
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
  }

  void stop() override { std::cout << "[" << name() << "] Stopping operator" << std::endl; }
};

// An operator that consumes data
class ConsumerOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(ConsumerOp)

  ConsumerOp() = default;

  void setup(holoscan::OperatorSpec& spec) override { spec.input<int>("in"); }

  void compute(holoscan::InputContext& op_input,
               [[maybe_unused]] holoscan::OutputContext& op_output,
               [[maybe_unused]] holoscan::ExecutionContext& context) override {
    auto data = op_input.receive<int>("in").value();

    // Consume the data
    std::cout << "[" << name() << "] Consuming data: " << data << std::endl;

    // Simulate some processing time
    std::this_thread::sleep_for(std::chrono::milliseconds(150));
  }

  void stop() override { std::cout << "[" << name() << "] Stopping operator" << std::endl; }
};

// A dedicated monitoring operator that runs independently of the main pipeline
class MonitorOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(MonitorOp)

  MonitorOp() = default;

  void setup(holoscan::OperatorSpec& spec) override {
    // No inputs or outputs - this operator runs independently
    spec.param(monitored_operators_,
               "monitored_operators",
               "Monitored Operators",
               "Names of operators to monitor",
               {});
  }

  void compute([[maybe_unused]] holoscan::InputContext& op_input,
               [[maybe_unused]] holoscan::OutputContext& op_output,
               [[maybe_unused]] holoscan::ExecutionContext& context) override {
    static int consecutive_idle_count = 0;

    std::cerr << "[" << name() << "] Operator status summary:" << std::endl;

    bool is_pipeline_idle = true;
    for (const auto& op_name : monitored_operators_.get()) {
      auto maybe_status = context.get_operator_status(op_name);
      if (!maybe_status) { throw std::runtime_error(maybe_status.error().what()); }
      auto status = maybe_status.value();
      std::cerr << "  - " << op_name << ": " << magic_enum::enum_name(status) << std::endl;
      if (status != holoscan::OperatorStatus::kIdle) { is_pipeline_idle = false; }
    }

    if (is_pipeline_idle) {
      consecutive_idle_count++;

      // Currently, there's no way to retrieve the computed SchedulingCondition status from the
      // operator (i.e., whether the computed scheduling condition status is NEVER).
      // Instead, we use `consecutive_idle_count` to determine if all operators, except the monitor
      // operator, have completed.
      // If `consecutive_idle_count` is equal to or greater than the hardcoded value 3,
      // we consider them completed, as there are cases where all operators are idle but not yet
      // completed.
      // A better approach for checking an operator's computed scheduling condition will be
      // available in a future release.
      if (consecutive_idle_count >= 3) {
        std::cerr << "[" << name() << "] All operators have completed." << std::endl;
        // Stop the monitor operator which is the only operator keeping the application alive
        stop_execution();  // the application will terminate through a deadlock
      }
    } else {
      consecutive_idle_count = 0;
    }
  }

 private:
  holoscan::Parameter<std::vector<std::string>> monitored_operators_;
};

class OperatorStatusTrackingApp : public holoscan::Application {
 public:
  void compose() override {
    using namespace holoscan;

    // Define the operators
    auto source = make_operator<FiniteSourceOp>(
        "source", make_condition<CountCondition>(10), Arg("max_count", 5));
    auto processor = make_operator<ProcessorOp>("processor");
    auto consumer = make_operator<ConsumerOp>("consumer");

    // Create the independent monitor operator with its own scheduling condition
    // This will continue running even if the main pipeline stops
    auto monitor = make_operator<MonitorOp>(
        "monitor",
        make_condition<PeriodicCondition>(std::chrono::milliseconds(50)),
        Arg("monitored_operators", std::vector<std::string>{"source", "processor", "consumer"}));

    // Define the workflow for the main pipeline
    add_flow(source, processor, {{"out", "in"}});
    add_flow(processor, consumer, {{"out", "in"}});

    // The monitor is not connected to any other operators - it runs independently
    add_operator(monitor);

    // Print information about the execution context API
    std::cout << "This example demonstrates the Operator Status Tracking API." << std::endl;
    std::cout << "The source operator will emit 5 values and then stop executing after 10 "
              << "iterations." << std::endl;
    std::cout << "The monitor operator runs independently and tracks the status of all operators."
              << std::endl;
    std::cout << "When operators complete, the monitor will be terminated." << std::endl;
    std::cout << "-------------------------------------------------------------------" << std::endl;
  }
};

int main() {
  auto app = holoscan::make_application<OperatorStatusTrackingApp>();
  holoscan::ArgList scheduler_args{holoscan::Arg("worker_thread_number", static_cast<int64_t>(2))};
  app->scheduler(app->make_scheduler<holoscan::EventBasedScheduler>("EBS", scheduler_args));
  app->run();

  std::cout << "-------------------------------------------------------------------" << std::endl;
  std::cout << "Application completed. All operators have finished processing." << std::endl;
  return 0;
}
